src/Sparrow.Server/ByteString.cs

```diff
@@ -562,93 +565,105 @@ public void LowMemoryOver()
+            // pool to only be half full on average. 
+            if (pool.Count >= SegmentPoolSize / 2 || pool.TryEnqueue(segment) == false)
+                segment.Dispose();
+                        
+            ALLOCATE:
+            // have to allocate it directly
             return new UnmanagedGlobalSegment(size);
         }
 
+        /// <summary>
+        /// Returns an UnmanagedGlobalSegment to the pool. If the segment is no longer in use and the pool is not full, 
+        /// it is added back for future reuse. 
+        /// 
+        /// If the pool is full, we attempt a fallback strategy:
+        /// 1. Dequeue another segment from the pool to compare sizes.
+        /// 2. Keep the larger segment (to better accommodate future allocations) and dispose of the smaller one.
+        /// 3. If re-enqueueing the chosen segment still fails, we dispose it to prevent memory leaks.
+        ///
+        /// This ensures that we maintain a healthy set of segments in the pool while also preventing indefinite growth or 
+        /// memory leaks if the pool cannot accept more segments.
+        /// </summary>
         public unsafe void Free(UnmanagedGlobalSegment memory)
         {
             ThrowIfNull<InvalidOperationException>(memory.Segment, "Attempt to return a memory segment that has already been disposed");
 
-            if (_minSize > memory.Size)
-            {
-                memory.Dispose();
-                return;
-            }
+            if (memory.InUse.Lower() == false)
+                ThrowSegmentStillInUse();
 
-            memory.InUse.Lower();
             memory.InPoolSince = DateTime.UtcNow;
 
-            var stack = SegmentsPool.Value;
+            var pool = SegmentsPool;
 
-            while (true)
+            // Attempt to return the segment to the pool.
+            if (pool.IsFull || pool.TryEnqueue(memory) == false)
             {
-                var current = stack.Head;
-                var newHead = new StackNode<UnmanagedGlobalSegment> { Value = memory, Next = current };
-                if (Interlocked.CompareExchange(ref stack.Head, newHead, current) == current)
-                    return;
+                // Pool is full; attempt to swap an existing segment with the current one
+                if (pool.TryDequeue(out var comparable))
+                {
+                    // Compare the sizes of the two segments.
+                    var (toAdd, toRemove) = comparable.Size < memory.Size
+                        ? (memory, comparable) // Keep the larger segment for reuse.
+                        : (comparable, memory); // Keep the existing segment if it's better.
+
+                    // Try to re-enqueue the chosen segment.
+                    if (pool.TryEnqueue(toAdd) == false)
+                        toAdd.Dispose(); // If enqueue fails, dispose to prevent leaks.
+
+                    // Dispose of the other segment.
+                    using (toRemove) ;
```

Comment:
Please check the warning here. I think you should do 

```
using (toRemove)
{ 
    // Try to re-enqueue the chosen segment.
    if (pool.TryEnqueue(toAdd) == false)
       toAdd.Dispose(); // If enqueue fails, dispose to prevent leaks.
}

Reply:
Makes sense even if `toAdd.Dispose()` fails we will call `toRemove.Dispose()` in the `finally` statement.